packages/web3-core/src/web3_request_manager.ts

```diff
@@ -219,42 +272,80 @@ export class Web3RequestManager<
 	private _processJsonRpcResponse<ResultType, ErrorType, RequestType>(
 		payload: JsonRpcPayload<RequestType>,
 		response: JsonRpcResponse<ResultType, ErrorType>,
+		{ legacy, error }: { legacy: boolean; error: boolean },
 	): JsonRpcResponse<ResultType> | never {
-		if (
-			jsonRpc.isBatchRequest(payload) &&
-			!Array.isArray(response) &&
-			!(response instanceof Error)
-		) {
-			throw new ResponseError(response, 'Got normal response for a batch request.');
+		// This is the majority of the cases so check these first
+		// A valid JSON-RPC response with error object
+		if (jsonRpc.isResponseWithError<ErrorType>(response)) {
+			throw new InvalidResponseError<ErrorType>(response);
 		}
 
-		if (!jsonRpc.isBatchRequest(payload) && Array.isArray(response)) {
-			throw new ResponseError(response, 'Got batch response for a normal request.');
+		// This is the majority of the cases so check these first
+		// A valid JSON-RPC response with result object
+		if (jsonRpc.isResponseWithResult<ResultType>(response)) {
+			return response;
 		}
 
-		if (jsonRpc.isBatchResponse(response)) {
+		if ((response as unknown) instanceof Error) {
+			throw response;
+		}
+
+		if (!legacy && jsonRpc.isBatchRequest(payload) && jsonRpc.isBatchResponse(response)) {
 			return response as JsonRpcBatchResponse<ResultType>;
 		}
 
+		if (legacy && !error && jsonRpc.isBatchRequest(payload)) {
+			return response as JsonRpcBatchResponse<ResultType>;
+		}
+
+		if (legacy && error && jsonRpc.isBatchRequest(payload)) {
+			// In case of error batch response we don't want to throw Invalid response
+			throw response;
+		}
+
 		if (
-			(jsonRpc.isResponseWithError(response) || jsonRpc.isResponseWithResult(response)) &&
-			!jsonRpc.isBatchRequest(payload)
+			legacy &&
+			!jsonRpc.isResponseWithError(response) &&
+			!jsonRpc.isResponseWithResult(response)
 		) {
-			if (response.id && payload.id !== response.id) {
-				throw new InvalidResponseError<ErrorType>(response);
+			const res = {
+				jsonrpc: '2.0',
+				// eslint-disable-next-line no-nested-ternary
+				id: jsonRpc.isBatchRequest(payload)
+					? payload[0].id
+					: 'id' in payload
+					? payload.id
+					: 'NON_EXISTING_ID',
```

Comment:
If there was an error in detecting the id in the Request object, it MUST be Null. instead of NON_EXISTING_ID 
 https://www.jsonrpc.org/specification

Reply:
https://www.jsonrpc.org/specification#id1
> The use of Null as a value for the id member in a Request object is discouraged, because this specification uses a value of Null for Responses with an unknown id. Also, because JSON-RPC 1.0 uses an id value of Null for Notifications this could cause confusion in handling.


This case is only for the legacy implementations which don't follow fully compliance JSON-RPC specification. If we use `null` the upper level code will fail as it checks the valid response must contains some id.

Reply:
[link you mentioned above ](https://www.jsonrpc.org/specification#id1) is for request object. and object under discussion is response. ( last line ) 

![image](https://user-images.githubusercontent.com/86780488/182352825-9d790f89-3d93-42fb-b364-797ce44fbdfc.png)

If upper level code in our lib always expects some id thn it should always put id in request obj for getting it in response.